### PR TITLE
add structured suggestions and fix false-positive for elided-lifetimes-in-paths lint

### PR DESCRIPTION
**Pull request by zackmdavis**
This adds structured suggestions to the elided-lifetimes-in-paths lint (introduced in Nov. 2017's #46254), prevents it from emitting a false-positive on anonymous (underscore) lifetimes (!), and adds it to the idioms-2018 group (#52041).

~~As an aside, "elided-lifetimes-in-paths" seems like an unfortunate name, because it's not clear exactly what "elided" means. The motivation for this lint (see original issue #45992, and [RFC 2115](https://github.com/rust-lang/rfcs/blob/e978a8d3017a01d632f916140c98802505cd1324/text/2115-argument-lifetimes.md#motivation)) seems to be specifically about not supplying angle-bracketed lifetime arguments to non-`&` types, but (1) the phrase "lifetime elision" has historically also referred to the ability to not supply a lifetime name to `&` references, and (2) an `is_elided` method in the HIR returns true for anoymous/underscore lifetimes, which is _not_ what we're trying to lint here. (That naming confusion is almost certainly what led to the false positive addressed here.) Given that the lint is relatively new and is allow-by-default, is it too late to rename it ... um, _again_ (#50879)?~~

~~This does _not_ address a couple of other false positives discovered in https://github.com/rust-lang/rust/issues/52041#issuecomment-402547901.~~

elided_states

r? @nikomatsakis 
cc @nrc @petrochenkov

### CODE REVIEW

**Comment by zackmdavis** [comment]
(Some would argue that we shouldn't give the lint the increased prominence of the idioms-2018 group (in contrast to allow-by-default, which I fear almost no one turns up) until we fix the false-positive exemplified by `Ref::map` (again, in https://github.com/rust-lang/rust/issues/52041#issuecomment-402547901 ), but the counterargument would be that anyone who enables the idioms-2018 lints before October should know that they're opting into alpha/beta software and might help us discover more edge cases?)

**Comment by nikomatsakis** [comment]
@zackmdavis 

I'm good with another name, do you have a suggestion? When it comes to the term "elided", there are two schools of thought in any case:

- "Elided" carries the plain English meaning, and hence refers to any lifetime is not written (my personal opinion)
- "Elided" refers speciically to a lifetime in a return type that is not written and is resolved to a lifetime from one of the parameters ("lifetime elision RFC")

Perhaps best to find another term altogether that is not as contentious. =)

It's sort of unclear what to call a `'_` lifetime, which is not explicitly written but also not invisible... ah, perhaps that is a good name?

`invisible_lifetimes_in_paths`?

**Comment by nikomatsakis** [comment]
> This does not address a couple of other false positives

Which ones are not addressed? It'd be nice to update the issue to have a current list of known false positives.

File: src/librustc/lint/builtin.rs [hunk 1]
```diff
@@ -252,7 +252,7 @@ declare_lint! {
 declare_lint! {
     pub ELIDED_LIFETIMES_IN_PATHS,
     Allow,
-    "hidden lifetime parameters are deprecated, try `Foo<'_>`"
+    "implicit lifetime parameters are deprecated"
```
Comment (nikomatsakis) [inline src/librustc/lint/builtin.rs]: As far as names go, my take:

- `implicit` is *ok*, but it tends to mean anything that the reader wants it to mean
- `hidden` and `invisible` I think are both ok

I don't think any name will necessarily convey perfectly what we are going for -- but all 3 names are roughly in the right direction. I would summarize as "a lifetime that you can't tell is there from looking at the type". For this, I like "hidden" or "invisible" better than "implicit".

File: src/librustc/middle/resolve_lifetime.rs [hunk 2]
```diff
@@ -2066,27 +2070,78 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
...
+        }
+
+        let mut err = self.tcx.struct_span_lint_node(
+            lint::builtin::ELIDED_LIFETIMES_IN_PATHS,
+            lifetime_refs[0].id, // FIXME: HirIdify #50928
+            segment_ident.span,
+            &format!("implicit lifetime parameters in types are deprecated"),
+        );
+
+        if num_implicit_lifetimes == 1 {
```
Comment (nikomatsakis) [inline src/librustc/middle/resolve_lifetime.rs]: Why only the case of 1 implicit lifetime?

File: src/librustc/middle/resolve_lifetime.rs [hunk 3]
```diff
@@ -2066,27 +2070,78 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
...
+            let (replace_span,
+                 suggestion) = if generic_args.args.len() == num_implicit_lifetimes &&
+                generic_args.bindings.is_empty() {
+                    // If there are no (non-implicit) generic args or bindings, our
+                    // suggestion includes the angle brackets
+                    (segment_ident.span.shrink_to_hi(), "<'_>")
+                } else {
+                    // Otherwise—sorry, this is kind of gross—we need to infer the
+                    // replacement point span from the generics that do exist
```
Comment (nikomatsakis) [inline src/librustc/middle/resolve_lifetime.rs]: I don't see a better way, but I get nervous around macros and things -- e.g., what will happen if somebody does:

`Ref<$x>`

in a macro? (I'm imagining that the span of `$x` might then come from some other place than the type itself...?) Maybe we want a test around that case? I think it'd be fine to disable suggestions there, but it'd be nice to ICE or suggest something mangled?

File: src/librustc/middle/resolve_lifetime.rs (hunk 2 above)
Comment (nikomatsakis) [reply src/librustc/middle/resolve_lifetime.rs]: It seems like the logic below would apply to any number of lifetimes, if we just generalized the string.

File: src/librustc/middle/resolve_lifetime.rs [hunk 4]
```diff
@@ -2066,27 +2070,78 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
...
+                    let mut first_generic_span = None;
+                    for ref arg in &generic_args.args {
+                        match arg {
+                            hir::GenericArg::Lifetime(lt) => {
```
Comment (nikomatsakis) [inline src/librustc/middle/resolve_lifetime.rs]: I think there cannot be a lifetime name here... you can **totally** elide lifetimes (e.g., `Foo<T>`) but you cannot write only **some** of the needed lifetimes (e.g., `Foo<'_, T>` when `Foo<'_, '_, T>` is needed) -- you'll get an error in the latter case and (probably?) not go down this path. Should test though.

File: src/test/ui/in-band-lifetimes/elided-lifetimes.rs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+// except according to those terms.
+
+// run-rustfix
+// compile-flags: --edition 2018
+
+#![allow(unused)]
+#![deny(elided_lifetimes_in_paths)]
+//~^ NOTE lint level defined here
+
+use std::cell::{RefCell, Ref};
```
Comment (nikomatsakis) [inline src/test/ui/in-band-lifetimes/elided-lifetimes.rs]: Can we test:

- [ ] some case with more than one parameter (just for fun, though I don't see why it wouldn't work)
- [ ] a case where there is more than one lifetime and you only supply *some*?
- [ ] the macro case where the type is constructed in a macro from bits and pieces?

**Review by nikomatsakis — APPROVED** [review]
Thank you very much for taking this up! Looks awesome. 💯 I have a few suggestions.

**Comment by zackmdavis** [comment]
@nikomatsakis 

(When you're [back from vacation](https://internals.rust-lang.org/t/nikos-availability-over-july-and-august/7904).)

> Why only the case of 1 implicit lifetime? It seems like the logic below would apply to any number of lifetimes, if we just generalized the string.

I couldn't find a situation where inserting `<'_, '_>` would result in successfully fixing broken code. As an illustrative negative example, consider this program—

```rust
#![allow(unused)]
#![warn(elided_lifetimes_in_paths)]

struct DoubleLifetime<'a, 'b> {
    one: &'a str,
    another: &'b str,
}

fn returns_double_lifetime(one: &'a str, another: &'b str) -> DoubleLifetime {
    DoubleLifetime { one, another }
}

fn main() {}
```

The lint does fire on the return type of `returns_double_lifetime`, but we also hit E0106 ("missing lifetime specifiers")—

```
warning: implicit lifetime parameters in types are deprecated
  --> elide.rs:17:63
   |
17 | fn returns_double_lifetime(one: &'a str, another: &'b str) -> DoubleLifetime {
   |                                                               ^^^^^^^^^^^^^^
   |

[...]

error[E0106]: missing lifetime specifiers
  --> elide.rs:17:63
   |
17 | fn returns_double_lifetime(one: &'a str, another: &'b str) -> DoubleLifetime {
   |                                                               ^^^^^^^^^^^^^^ expected 2 lifetime parameters
```

Adding `<'_, '_>` doesn't solve the E0106—

```
error[E0106]: missing lifetime specifiers
  --> elide.rs:17:78
   |
17 | fn returns_double_lifetime(one: &'a str, another: &'b str) -> DoubleLifetime<'_, '_> {
   |                                                                              ^^ expected 2 lifetime parameters
```

I think this makes sense—if there are two distinct "output" lifetimes (in [the terminology of RFC 141](https://github.com/rust-lang/rfcs/blob/master/text/0141-lifetime-elision.md#lifetime-positions)), they can't both be anonymous, because they need names to disambiguate which is which, right? And this is a different scenario from when we can legitimately use multiple anonymous _input_ lifetimes `<'_, '_>` in an impl header [as illustrated in the Edition Guide](https://rust-lang-nursery.github.io/edition-guide/2018/transitioning/ownership-and-lifetimes/anonymous-lifetime.html#more-details)?

> Which ones are not addressed? It'd be nice to update the issue to have a current list of known false positives.

I think I've fixed them now. What _was_ happening is that the lint would fire on the type path in struct construction expressions—

```
error: implicit lifetime parameters in types are deprecated
  --> src/test/ui/in-band-lifetimes/elided-lifetimes.rs:44:5
   |
44 |     WrappedWithBow { gift }
   |     ^^^^^^^^^^^^^^- help: indicate the anonymous lifetime: `<'_>`
```

—and in associated-functions/static-methods (the motivating examples being `Ref::clone` and `Ref::map`)—

```
warning: implicit lifetime parameters in types are deprecated
  --> elide.rs:20:15
   |
20 |     let b2a = Ref::clone(b1);
   |               ^^^-^^^^^^
   |                  |
   |                  help: indicate the anonymous lifetime: `<'_>`
```

In both these cases, applying the suggestion results in code that fails to even parse ("expected `:`, found `>`").

To avoid this, I first moved the lint check from (a callee of) the `visit_path` method in the `LifetimeContext` visitor, to the `hir::TyPath` arm of `visit_ty`. That in itself fixes the struct-constructor case, but we were still emitting a false-positive in `Ref::clone`-like cases. (As far as the HIR is concerned, it seems that `Ref::clone` is the same as the UFCS expression `<Ref>::clone`, where the `Ref` counts as a type to be visited by `visit_ty`.) To fix _that_, I'm afraid I couldn't think of anything better than [just not letting the visitor recurse into the callee of call expressions](https://github.com/rust-lang/rust/commit/4c31c07c4f1b438c5ac48b50ee5202139a1aaf4b). I _think_ that's legitimate, because I'm _pretty sure_ you can't have a named lifetime in need of resolution there?—but **this is in particular need of your expert reviewer scrutiny** before we merge it. ( @oli-obk might also have insight, as the author of [the other work](https://github.com/rust-lang/rust/blame/49f1e5d8ab6edc4249752a0ee53b771d46d8ac49/src/librustc/middle/resolve_lifetime.rs#L628) we're doing in the `TyPath` arm.)

> I think there cannot be a lifetime name here... you can totally elide lifetimes (e.g., Foo<T>) but you cannot write only some of the needed lifetimes (e.g., Foo<'_, T> when Foo<'_, '_, T> is needed) -- you'll get an error in the latter case and (probably?) not go down this path. Should test though.

Thanks for clarifying this. However, unless I can convince myself that we error out _before_ we get here, I think I'm inclined to naïvely obey the type system rather than asserting that the `Lifetime` arm never happens, on separation-of-concerns grounds: this method is _just_ about linting for invisible lifetimes in otherwise (currently!) legal code; erroring out is someone else's responsibility (I added [a comment](https://github.com/zackmdavis/rust/blob/a9fa345dadb18e7cf8dfe6e718bef606fce070f1/src/librustc/middle/resolve_lifetime.rs#L2104-L2106).)

> It's sort of unclear what to call a '_ lifetime, which is not explicitly written but also not invisible...

The [edition guide calls it](https://rust-lang-nursery.github.io/edition-guide/2018/transitioning/ownership-and-lifetimes/anonymous-lifetime.html) an "anonymous" lifetime, which seems appropriate (it's not omitted from the source, it just doesn't have a name). I've followed this terminology in the span label here ("indicate the anonymous lifetime").

> I'm good with another name [...] `invisible_lifetimes_in_paths`? `hidden_lifetime_in_paths`?

After re-consulting [RFC 141](https://github.com/rust-lang/rfcs/blob/master/text/0141-lifetime-elision.md), I'm **not enthusiastic about renaming the lint anymore**.

(The reason I thought we wanted to rename is because I had mentally associated the word "elision" with `&` references specifically, which this lint isn't concerned with, but the RFC makes it clear that unwritten angle-bracked lifetimes are also "elided", and on reflection, I think "in paths" does a good enough job of being more specific.)

**Comment by zackmdavis** [comment]
(CI error is mysterious; this PR does not touch librustc_metadata)

**Comment by oli-obk** [comment]
You definitely broke something, the question is whether this code wasn't broken to begin with:

```rust
        let registrar = unsafe {
            let sym = match lib.symbol(&sym) {
                Ok(f) => f,
                Err(err) => self.sess.span_fatal(span, &err),
            };
            mem::transmute::<*mut u8, fn(&mut dyn Registry)>(sym)
        };
```

**Comment by zackmdavis** [comment]
Thanks, my code is almost certainly doing the wrong thing on `mem::transmute::<*mut u8, fn(&mut dyn Registry)>(sym)`. 

I'll need to rethink this (maybe we can store some extra state on the `LifetimeContext` visitor, rather than mucking with its traversal, which I _really_ should have known better than to think I could get away with), but don't have time today.

**Comment by nikomatsakis** [comment]
@zackmdavis 

> I couldn't find a situation where inserting <'_, '_> would result in successfully fixing broken code. 

Perhaps this?

```rust
#![allow(unused)]
#![warn(elided_lifetimes_in_paths)]

struct DoubleLifetime<'a, 'b> {
    one: &'a str,
    another: &'b str,
}

fn accepts_double_lifetime(dl: DoubleLifetime) -> String {
    format!("{} {}", dl.one, dl.another)
}

fn main() {}
```

> As far as the HIR is concerned, it seems that `Ref::clone` is the same as the UFCS expression `<Ref>::clone`, where the `Ref` counts as a type to be visited by `visit_ty`.

Hmm, yes, I can imagine that. I wonder if we are doing this lint in the wrong place. Maybe we should do it during **HIR lowering**...? After all, that is where we actually insert the various markers in HIR in the first place. I forget why I didn't suggest that route.

**Comment by nikomatsakis** [comment]
> I think that's legitimate, because I'm pretty sure you can't have a named lifetime in need of resolution there?—but this is in particular need of your expert reviewer scrutiny before we merge it.

I would say it's not the worst thing ever, but it seems suboptimal. My goal is roughly to make lifetime parameters work the same as types -- that is, whenever you would have to write the type parameters for `Foo`, you also write the lifetime ones. So if you would do `Vec<_>` (and not just `Vec`) then you should do `Foo<'_>` (and not just `Foo`).

**Comment by zackmdavis** [comment]
> Maybe we should do it during HIR lowering...?

(This [is looking like](https://github.com/zackmdavis/rust/commit/4f51df67ac61c) it's going to be _much_ better, but for now my build is mysteriously ICEing and I need to sleep.)

**Comment by zackmdavis** [comment]
@nikomatsakis ready :checkered_flag:

**Review by nikomatsakis — APPROVED** [review]

**Comment by kennytm** [comment]
Caused ICE with `dist-x86_64-msvc-alt` during `BuiltinLintDiagnostics::run()`.

```
   Compiling syn v0.14.4
thread 'main' panicked at 'byte index 89352 is out of bounds of `$crate::synom::PResult<$o>`', libcore\str\mod.rs:2094:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

<details><summary>The stack trace from the debugger:</summary>

```
0000008c`bc9739d8 00007ffd`22a913ed ntdll!NtWaitForMultipleObjects+0xa
0000008c`bc9739e0 00007ffd`23b67d51 KERNELBASE!WaitForMultipleObjectsEx+0xed
0000008c`bc973cc0 00007ffd`23b67773 kernel32!WerpReportFaultInternal+0x581
0000008c`bc974230 00007ffd`22b71c1f kernel32!WerpReportFault+0x83
0000008c`bc974260 00007ffd`258ff1b3 KERNELBASE!UnhandledExceptionFilter+0x23f
0000008c`bc974350 00007ffd`258e1e26 ntdll!RtlUserThreadStart$filt$0+0x3e
0000008c`bc974390 00007ffd`258f349d ntdll!_C_specific_handler+0x96
0000008c`bc974400 00007ffd`258b48d7 ntdll!RtlpExecuteHandlerForException+0xd
0000008c`bc974430 00007ffd`258f262a ntdll!RtlDispatchException+0x197
0000008c`bc974b00 00007ffd`22b6df12 ntdll!KiUserExceptionDispatch+0x3a
*** WARNING: Unable to verify checksum for rustc-158b3752b6ee1be6.dll
0000008c`bc975238 00007ffd`0727c83a KERNELBASE!DebugBreak+0x2
*** WARNING: Unable to verify checksum for std-822da15d609da1ff.dll
0000008c`bc975240 00007ffd`0ca68ca2 rustc_158b3752b6ee1be6!ZN5rustc4util6common10panic_hook17h7a7f3dae1f9492b5E+0x19a
0000008c`bc9752e0 00007ffd`0ca68804 std_822da15d609da1ff!std::panicking::rust_panic_with_hook+0x1a2
0000008c`bc975390 00007ffd`0ca686e9 std_822da15d609da1ff!std::panicking::continue_panic_fmt+0xa4
0000008c`bc975430 00007ffd`0ca9eedc std_822da15d609da1ff!std::panicking::rust_begin_panic+0x9
0000008c`bc975460 00007ffd`0cab1e7b std_822da15d609da1ff!core::panicking::panic_fmt+0x3c
0000008c`bc9754c0 00007ffd`07177a58 std_822da15d609da1ff!core::str::slice_error_fail+0x2fb
0000008c`bc9755d0 00007ffd`072d1314 rustc_158b3752b6ee1be6!ZN5rustc4lint7builtin22BuiltinLintDiagnostics3run17h4bb3542f622712c3E+0xb38
0000008c`bc9757a0 00007ffd`072dc943 rustc_158b3752b6ee1be6!ZN5rustc4lint7context12EarlyContext8check_id17h7f0064d9b036bf15E+0x4a4
*** WARNING: Unable to verify checksum for rustc_driver-b7c23ebe756beb8b.dll
0000008c`bc975b30 00007ffd`1a82e309 rustc_158b3752b6ee1be6!ZN5rustc4lint7context15check_ast_crate17hf056b9ed4f51c850E+0x3f3
0000008c`bc975e60 00007ffd`1a827b4c rustc_driver_b7c23ebe756beb8b!ZN12rustc_driver6driver11count_nodes17h72939dcc6d63838eE+0x2339
0000008c`bc977c90 00007ffd`1a8c8099 rustc_driver_b7c23ebe756beb8b!ZN12rustc_driver6driver13compile_input17h7b701af89d68eeaeE+0x65c
0000008c`bc97a010 00007ffd`1a82698c rustc_driver_b7c23ebe756beb8b!ZN12rustc_driver22run_compiler_with_pool17h1e7dab3c08524b6dE+0x1009
0000008c`bc97bbf0 00007ffd`1a7db3fb rustc_driver_b7c23ebe756beb8b!ZN6syntax12with_globals17h3b145e5e099dcffdE+0x60c
0000008c`bc97d190 00007ffd`0ca9a732 rustc_driver_b7c23ebe756beb8b!ZN101_$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$9call_once17h66b0d8204f471548E+0xbb
0000008c`bc97f9e0 00007ffd`1a8c43e7 std_822da15d609da1ff!panic_unwind::__rust_maybe_catch_panic+0x22
0000008c`bc97fa40 00007ffd`1a8d664e rustc_driver_b7c23ebe756beb8b!ZN12rustc_driver7profile22profile_queries_thread17heb501a38eb0cafebE+0x5807
*** WARNING: Unable to verify checksum for rustc.exe
0000008c`bc97fd10 00007ff6`3ea71006 rustc_driver_b7c23ebe756beb8b!ZN12rustc_driver4main17h055a7d7a4cdec127E+0xe
0000008c`bc97fd40 00007ffd`0ca68647 rustc!__acrt_rg_language_count <PERF> (rustc+0x1006)
(Inline Function) --------`-------- std_822da15d609da1ff!std::rt::lang_start_internal::{{closure}}+0xc
0000008c`bc97fd70 00007ffd`0ca9a732 std_822da15d609da1ff!std::panicking::try::do_call<closure,i32>+0x17
0000008c`bc97fda0 00007ffd`0ca65133 std_822da15d609da1ff!panic_unwind::__rust_maybe_catch_panic+0x22
(Inline Function) --------`-------- std_822da15d609da1ff!std::panicking::try+0x33
(Inline Function) --------`-------- std_822da15d609da1ff!std::panic::catch_unwind+0x33
0000008c`bc97fe00 00007ff6`3ea7107a std_822da15d609da1ff!std::rt::lang_start_internal+0xa3
0000008c`bc97fea0 00007ff6`3ea71299 rustc!main+0x2a
(Inline Function) --------`-------- rustc!invoke_main+0x22
0000008c`bc97fed0 00007ffd`23a613d2 rustc!__scrt_common_main_seh+0x11d
0000008c`bc97ff10 00007ffd`258754f4 kernel32!BaseThreadInitThunk+0x22
0000008c`bc97ff40 00000000`00000000 ntdll!RtlUserThreadStart+0x34
```

</details>

**Comment by zackmdavis** [comment]
Hm, the `snippet.split_at((insertion_span.lo().0 - path_span.lo().0) as usize);` is probably not legitimate in the presence of macros. I'll take a closer look soon.

**Comment by zackmdavis** [comment]
I added [a check](https://github.com/rust-lang/rust/blob/53cebfde949f42ca0d406b42520032017e123a35/src/librustc/lint/builtin.rs#L457-L462) that I'm quite confident will avoid the panic we saw in the AppVeyor Windows build. This _does_ imply that there exist some cases with macros for which we'll issue no suggestion (or possibly a wrong suggestion), but I don't think this should be a blocking concern. (Note that the UI test [does demonstrate some simple macro scenarios](https://github.com/rust-lang/rust/blob/53cebfde949f42ca0d406b42520032017e123a35/src/test/ui/in-band-lifetimes/elided-lifetimes.stderr#L37-L53) that we handle correctly.) Thus, I argue that this should be re-r+'d.

@nikomatsakis :checkered_flag:

**Comment by nikomatsakis** [comment]
@zackmdavis great!

**Comment by nikomatsakis** [comment]
@bors r+ p=1

This is important for EP2

File: src/librustc/lint/builtin.rs [hunk 6]
```diff
@@ -442,6 +443,36 @@ impl BuiltinLintDiagnostics {
...
+                n, path_span, incl_angl_brckt, insertion_span, anon_lts
+            ) => {
+                let (replace_span, suggestion) = if incl_angl_brckt {
+                    (insertion_span, anon_lts)
+                } else {
+                    // When possible, prefer a suggestion that replaces the whole
+                    // `Path<T>` expression with `Path<'_, T>`, rather than inserting `'_, `
+                    // at a point (which makes for an ugly/confusing label)
+                    if let Ok(snippet) = sess.codemap().span_to_snippet(path_span) {
+                        let insertion_index = (insertion_span.lo().0 - path_span.lo().0) as usize;
```
Comment (kennytm) [inline src/librustc/lint/builtin.rs]:
```
   Compiling termcolor v0.3.6
thread 'main' panicked at 'attempt to subtract with overflow', librustc\lint\builtin.rs:456:47
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: Could not compile `serde_derive`.
```

😂

**Comment by zackmdavis** [comment]
@kennytm @nikomatsakis I forgot that integer overflow panics (in non-release builds). Should be fixed now.

**Comment by oli-obk** [comment]
I think we have `#[deny(warnings)]`

**Pull request merged**